### PR TITLE
configure.py: don't modify `modes` in write_build_file()

### PR DESCRIPTION
**Pull request by michoecho**
The true motivation for this patch is a certain problem with configure.py in scylla-enterprise, which can only be solved by moving the `extra_cxxflags` lines before configure_seastar(). This patch does that by hoisting get_extra_cxxflags() up to create_build_system().

But this patch makes sense even if we disregard the real motivation. It's weird that a function called `write_build_file()` adds additional build flags on its own.

### CODE REVIEW

**Review by avikivity — APPROVED** [review]

File: configure.py
Comment (tchaikov) [inline configure.py:2410]: does it make sense to only update the `build_modes` which is a subset of `mode`. this subset contains the modes which are enabled by user with `--mode` option. this would be more consistent with the change you are adding for querying seastar's exposed building options. also, could use something like,
```py
for mode, config in build_modes.items():
    extra_cxxflags = ' '.join(get_extra_cxxflags(mode, config, args.cxx, args.debuginfo))
    # likewise
```

File: configure.py
Comment (tchaikov) [inline configure.py:2423]: ditto.

**Review by tchaikov — APPROVED** [review]
thank you for the fix. and sorry for introducing this regression.

this change lgtm apart from some tiny nits. they are far from blockers. i will send follow-up changes to implement them, if they will not be addressed in this PR.

File: configure.py
Comment (michoecho) [reply configure.py]: Done.

File: configure.py
Comment (michoecho) [reply configure.py]: Done.

**Comment by michoecho** [comment]
> thank you for the fix. and sorry for introducing this regression.

Note: it's not a fix yet. The fix requires an additional change in scylla-enterprise's configure.py, after this is merged there.

**Comment by michoecho** [comment]
The failed test is unrelated.

**Comment by tchaikov** [comment]
> #### 🔴 CI State: FAILURE
> 
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6521/) ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6521/) ❌ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7760/) ✅ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6521/)
> #### Failed Tests (1/24453):
> 
>     * [test_simple_kill_node_while_decommissioning](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6521/testReport/junit/update_cluster_layout_tests/TestUpdateClusterLayout/Tests___dtest___test_simple_kill_node_while_decommissioning) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_simple_kill_node_while_decommissioning)
> 
> 
> #### Build Details:
> 
>     * Duration: 1 hr 40 min
> 
>     * Builder: spider6.cloudius-systems.com

looks like #17124

**Comment by denesb** [comment]
Re-kicked CI.

**Comment by michoecho** [comment]
Can't we just ignore the failed CI if the failure is obviously irrelevant? Why waste CI cycles?

**Comment by mykaul** [comment]
> Can't we just ignore the failed CI if the failure is obviously irrelevant? Why waste CI cycles?

Ignoring CI failures leads to the dark side.
But how do we know test failures are irrelevant? And if we do know, could we have skipped them in the 1st place?

(But practically, I do feel we could make a better judgement call - if we have known CI failures, then I'd like to trust the maintainers to be able to merge if they agree the failure is irrelevant)

**Comment by michoecho** [comment]
> > Can't we just ignore the failed CI if the failure is obviously irrelevant? Why waste CI cycles?
> 
> Ignoring CI failures leads to the dark side. But how do we know test failures are irrelevant? And if we do know, could we have skipped them in the 1st place?
> 
> (But practically, I do feel we could make a better judgement call - if we have known CI failures, then I'd like to trust the maintainers to be able to merge if they agree the failure is irrelevant)

You answered your own question — if test was observed to fail even without the PR, then the failure is probably not caused by the PR.

Besides, just kicking CI for a PR until it passes does not protect you from regressions any more than just pushing the PR to `next`. If the failure is reliable, next will hit it. If it isn't, the regression will get kicked through the CI anyway.

**Comment by mykaul** [comment]
In theory a failed test could be flaky and even known flaky, but could be due to your patch, so the failure needs to be looked at. In practice, it's quite rare. Which is why we need to focus on both fixing them and understand if they are a flaky one - the latter is being worked on.

**Pull request closed without merging**
